### PR TITLE
Guidelines on Licensing and DCO for Radicle

### PR DESCRIPTION
**Pull request by FintanH**
[Rendered](https://github.com/radicle-dev/radicle-decisions/blob/0003-licensing/proposals/0003.md)

### CODE REVIEW

File: proposals/0003.md
Comment (kim) [inline proposals/0003.md:32]: The license file actually needs to include the linking exception text (at the top?).

**Review by kim — COMMENTED** [review]
I also think that it is at least _recommended_, if not _required_, to put a license header in each file which is part of the final distribution. I am, however, hoping that we might be able to use a shorter header, like [here](https://github.com/libgit2/libgit2/blob/master/src/alloc.c#L1-L6)

**Comment by FintanH** [comment]
> I also think that it is at least recommended, if not required, to put a license header in each file which is part of the final distribution. I am, however, hoping that we might be able to use a shorter header, like here

Let's just make it required and be on the safe side. How can we confirm if we can use the shorter header? It seems like it covers the bases with the pointer to `COPYING`.

File: proposals/0003.md
Comment (CodeSandwich) [reply proposals/0003.md:32]: ~There are also SPDX headers, which indicate the license in a single line. Not sure about the legal might of these though~ SPDX doesn't support our license flavor

File: proposals/0003.md
Comment (CodeSandwich) [inline proposals/0003.md]: No. https://github.com/isaacs/github/issues/5 Somebody posted there a workaround, that by protecting the branches with impossible rules one can effectively turn the green button of doom into a red button of nuclear wasteland.

As far as I know GitHub doesn't support `ff-only` merging of PRs, there's now way to prevent it from adding its commits.

File: proposals/0003.md
Comment (CodeSandwich) [inline proposals/0003.md]: I think that there's one more problem worth thinking about: how do we enforce that the CI has passed when merging into master?

File: assets/LICENSE
Comment (CodeSandwich) [inline assets/LICENSE]: In the smart contracts we're using the SPDX license headers. I wonder if [SPDX-License-Identifier: (GPL-3.0-only WITH GPL-3.0-linking-exception)](https://spdx.org/licenses/GPL-3.0-linking-exception.html) will be compatible with this license?

File: proposals/0003.md
Comment (rudolfs) [inline proposals/0003.md]:
```suggestion
We SHALL NOT merge pull requests using the GitHub UI. This is due to GitHub
```

File: assets/LICENSE
Comment (FintanH) [reply assets/LICENSE]: I'm going to ping Ange, since she's the one advising me on the legalese

Comment (CodeSandwich) [reply assets/LICENSE]: Thanks! If it's good, we could use it in all the repos, it's much lighter than our custom header

Comment (pbjam12345) [reply assets/LICENSE]: Andrew mentioned in [his post](https://radicle.community/t/radicle-licensing-model/282), but with regards to the Radicle Linking Exception: "Unfortunately, there is no current SPDX identifier for this exception, but we are looking to apply for one from SPDX." So while it's heavier, let's use our custom header at this time - though we will hopefully be able to go lighter in the future.

File: assets/license-header
Comment (NunoAlexandre) [inline assets/license-header]: Regarding the `2020`, is this something we will need to bump every year? as in, will we have a PR on every Jan 1 updating it to the new year?

File: assets/license-header
Comment (NunoAlexandre) [inline assets/license-header]: is there a reason as to why this line is not wrapped like all the others?

File: assets/license-headers-check
Comment (NunoAlexandre) [inline assets/license-headers-check:3]: Shouldn't we make htis work for all source files we work with? `*.rs`, `*.js`, `*.svelte`, `*.ts`, ~~*.hs~~ #criesinpain, etc?

File: assets/license-header
Comment (FintanH) [reply assets/license-header]: cc @angedupre

File: assets/license-header
Comment (FintanH) [reply assets/license-header]: I left it unformatted, but can wrap it too

File: assets/license-headers-check
Comment (FintanH) [reply assets/license-headers-check:3]: Yes please, contributions welcome :smile:

File: proposals/0003.md
Comment (geigerzaehler) [inline proposals/0003.md]: What does this mean? Are you referring to the “Merge Button” setting of a Github repo or a branch protection rule? Does this even apply if we don’t merge via the Github UI?

Comment (FintanH) [reply proposals/0003.md]: Ya, so we don't merge via the button. But I was thinking in the case that somebody _does_ press the button then if it's a rebase then there's no extra commit.

Comment (geigerzaehler) [reply proposals/0003.md]: So you’re referring to the “Allow rebase merging” “Merge Button” setting. Even with that the history will not be made up of signed commits because Github will create new commits for every commit in the PR.

File: proposals/0003.md
Comment (geigerzaehler) [inline proposals/0003.md]: I’m wondering whether we want to specify the actual process for this in this document. For example whether we use merge commits are fast-forwards and who pushes the changes to master. Alternatively we can do this later in the different repos.

File: proposals/0003.md
Comment (CodeSandwich) [reply proposals/0003.md]: I've tested it and the "rebase merge" on GitHub is just broken. Even if you merge a fast-forward branch, it will still rebase it with its dirty unsigned fingers.

File: assets/license-header
Comment (kim) [inline assets/license-header]: omg seriously? All the information needs to be included in _every_ source file?

File: proposals/0003.md
Comment (FintanH) [reply proposals/0003.md]: Ach! Thanks for checking. One other idea is to require 6 maintainers to approve which we rarely get :joy:

File: assets/license-header
Comment (FintanH) [reply assets/license-header]: Maybe not. Checking with the legals :sweat_smile:

File: proposals/0003.md
Comment (kim) [inline proposals/0003.md]: Could we change terminology to `signed-off` vs `signed`, where `signed` means that it is _cryptographically_ signed?

File: proposals/0003.md
Comment (kim) [reply proposals/0003.md]: I think this section is a bit inaccurate: technically, a merge commit which does not alter the `tree` does not bear any copyrightable contribution _per definitionem_, and thus does not need a sign-off. It is, however, not entirely clear if THE GREEN BUTTON always does the right thing wrt to `Signed-off-by`, especially when the branch to be merged contains commits by multiple authors. Therefore, not using THE GREEN BUTTON is preferable.

I do **not** think we actually have to require that all commits be _cryptographically_ signed, only release tags (but that is a separate discussion).

Not wanting commits made by GitHub (which, coincidentally, _are_ cryptographically signed) is an entirely separate issue, and only a matter of policy.

File: proposals/0003.md
Comment (kim) [reply proposals/0003.md]: As per my other [comment](https://github.com/radicle-dev/radicle-decisions/pull/5/files#r531633613), I am not against allowing squash commits via THE GREEN BUTTON. There, the button-operator will have the opportunity to review the commit message (just make sure you don't add a `Signed-off-by` on behalf of someone else lol).

The only reason to not use THE GREEN BUTTON at all is not wanting GitHub as a contributor.

Comment (kim) [reply proposals/0003.md]: Oh wait, scrap that. A squash merge _MUST_ be signed-off. As it's not you, but GitHub, who introduces a _new_ `tree`, you would need to add _your_ `Signed-off-by` in the textbox. At this point, the CLI will win in terms of convenience.

File: assets/license-header
Comment (FintanH) [reply assets/license-header]: Ok, so @angedupre is going to give us the copy that doesn't include all of this and looks more like the `libgit2` one you shared earlier :+1:

**Comment by kim** [comment]
Following radicle-dev/radicle-avatar#2, I think it was also recommended that "Shared Libraries" (definition: doesn't implement aspects of the protocol) be licensed permissively. The resoning is that the increased simplicity for everyone involved trumps uniformity.

Following conventions in the Rust community, I'd say this would be `Apache-2.0 OR MIT`. Perhaps we should include this here, although we may punt on the procedural question of deciding whether or not a new library qualifies.

**Comment by cloudhead** [comment]
> Following conventions in the Rust community, I'd say this would be `Apache-2.0 OR MIT`. Perhaps we should include this here, although we may punt on the procedural question of deciding whether or not a new library qualifies.

Does this mean we need the linking exception etc. in both license files?

File: assets/license-header
Comment (pbjam12345) [reply assets/license-header]: If you can do that, that sounds like great plan :)

Comment (FintanH) [reply assets/license-header]: I'll leave a note in the recommendations about this. I guess we should have a script to do this as well.

File: assets/LICENSE
Comment (pbjam12345) [reply assets/LICENSE]: @CodeSandwich Sorry, I may have misunderstood your comment here! Yes, while we can't use an SPDX identifier at this time, we don't need to use the WHOLE wrapper about the Radicle Linking Exception in each source file :)

**Comment by kim** [comment]
@cloudhead We don't need a linking exception for `Apache-2.0 OR MIT` -- it is already granted by the license. We _can_ have the DCO -- after all, it just says that the contributor is allowed to contribute under the license -- but tbh, I wouldn't bother until I'd see a _really substantial_ contribution where I would want to ask (we're talking about `radicle-avatar` atm).

File: assets/license-header
Comment (pbjam12345) [reply assets/license-header]: Yep! Created some condensed copy for @FintanH for inclusion at the top of the source files. As mentioned in a comment above - there was some confusion on my side but I think we're clear now and have copy that makes sense.

**Comment by angedupre** [comment]
@kim Why not have a DCO in the permissively licensed repos for consistency? Agreed that the Radicle Linking Exception is only to be included with those repos licensed under GPLv3.

**Comment by pbjam12345** [comment]
@kim Why not have a DCO in the permissively licensed repos for consistency? Agreed that the Radicle Linking Exception is only to be included with those repos licensed under GPLv3.

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]:
```suggestion
The decision here is to implement most of the Radicle software under the license of GPLv3 + Radicle Linking Exception. We are also asking outside contributors to sign-off on their commits to our software, as described by our Developer Certificate of Origin (DCO).
```

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md:40]: @cloudhead Are we moving forward with MIT only?

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]:
```suggestion
## Copyright and License Header
```

**Comment by cloudhead** [comment]
> @cloudhead We don't need a linking exception for `Apache-2.0 OR MIT` -- it is already granted by the license.

Hah of course :sweat_smile: 

>  but tbh, I wouldn't bother until I'd see a really substantial contribution

Yeah I agree..

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]:
```suggestion
If you've found you've pushed a few commits without signed-off-by, you can still add it afterwards.
```

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]:
```suggestion
## Copyright and License Header

For consistency and clarity, we must include a header at the top of each
source file. The template for this header can be found
[here](./path/to/license-header). If you are creating a new file in the
repository you will have to add this header to the file.
```

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]:
```suggestion
While it is not strictly necessary, we are RECOMMENDING that maintainers
from Monadic sign-off on their commits as well. The easiest way to set this is
up by including it a `git commit` alias, for example:
```

File: proposals/0003.md
Comment (pbjam12345) [inline proposals/0003.md]: Using our @monadic.xyz isn't absolutely critical - but it makes things more clear and consistent from a legal perspective.

File: assets/license-header
Comment (kim) [reply assets/license-header]: Acc. to [Wikipedia](https://en.wikipedia.org/wiki/Copyright_notice#Form_of_notice_for_visually_perceptible_copies), US law says it should be the year of first publication. So, unless there are major copyrightable modifications, it doesn't have to be updated. And vice versa: if the file gets substantially modified, add another year, like `1996, 2032`.

**Review by kim — APPROVED** [review]
Modulo editorials, lgtm

**Review by geigerzaehler — APPROVED** [review]

**Review by rudolfs — APPROVED** [review]

**Review by CodeSandwich — APPROVED** [review]

**Review by xla — APPROVED** [review]
🈚️ 🔝 ☢ 🏫

**Comment by kim** [comment]
Closed via 250b03d02a5c57a2b9467d3430d76b48d01871f9

**Pull request closed without merging**
